src/GafferImage/CollectImages.cpp

```diff
@@ -0,0 +1,447 @@
+	ImageProcessor::affects( input, outputs );
+
+	const ImagePlug *imagePlug = input->parent<ImagePlug>();
+	if( imagePlug && imagePlug == inPlug() )
+	{
+		if( input == imagePlug->dataWindowPlug() )
+		{
+			outputs.push_back( outPlug()->dataWindowPlug() );
+		}
+
+		if( input == imagePlug->channelNamesPlug() )
+		{
+			outputs.push_back( outPlug()->channelNamesPlug() );
+		}
+
+		if( input == imagePlug->channelDataPlug() )
+		{
+			outputs.push_back( outPlug()->channelDataPlug() );
+		}
+
+		if( input == imagePlug->formatPlug() )
+		{
+			outputs.push_back( outPlug()->formatPlug() );
+		}
+
+		if( input == imagePlug->metadataPlug() )
+		{
+			outputs.push_back( outPlug()->metadataPlug() );
+		}
+	}
+	else if( input == layerNamesPlug() || input == layerContextVariablePlug() )
+	{
+		outputs.push_back( outPlug()->channelNamesPlug() );
+		outputs.push_back( outPlug()->channelDataPlug() );
+		outputs.push_back( outPlug()->dataWindowPlug() );
+		outputs.push_back( outPlug()->formatPlug() );
+		outputs.push_back( outPlug()->metadataPlug() );
+	}
+
+}
+
+void CollectImages::hashFormat( const GafferImage::ImagePlug *parent, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ImageProcessor::hashFormat( parent, context, h );
+	ConstStringVectorDataPtr layerNamesData = layerNamesPlug()->getValue();
+
+	if( layerNamesData->readable().size() )
+	{
+		Context::EditableScope editScope( context );
+		editScope.set( layerContextVariablePlug()->getValue(), layerNamesData->readable()[0] );
+		inPlug()->formatPlug()->hash( h );
+	}
+}
+
+GafferImage::Format CollectImages::computeFormat( const Gaffer::Context *context, const ImagePlug *parent ) const
+{
+	ConstStringVectorDataPtr layerNamesData = layerNamesPlug()->getValue();
+
+	if( layerNamesData->readable().size() )
+	{
+		Context::EditableScope editScope( context );
+		editScope.set( layerContextVariablePlug()->getValue(), layerNamesData->readable()[0] );
+		return inPlug()->formatPlug()->getValue();
+	}
+	else
+	{
+		return outPlug()->formatPlug()->defaultValue();
+	}
+}
+
+void CollectImages::hashMetadata( const GafferImage::ImagePlug *parent, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ImageProcessor::hashMetadata( parent, context, h );
+	ConstStringVectorDataPtr layerNamesData = layerNamesPlug()->getValue();
+
+	if( layerNamesData->readable().size() )
+	{
+		Context::EditableScope editScope( context );
+		editScope.set( layerContextVariablePlug()->getValue(), layerNamesData->readable()[0] );
+		inPlug()->metadataPlug()->hash( h );
+	}
+}
+
+IECore::ConstCompoundDataPtr CollectImages::computeMetadata( const Gaffer::Context *context, const ImagePlug *parent ) const
+{
+	ConstStringVectorDataPtr layerNamesData = layerNamesPlug()->getValue();
+
+	if( layerNamesData->readable().size() )
+	{
+		Context::EditableScope editScope( context );
+		editScope.set( layerContextVariablePlug()->getValue(), layerNamesData->readable()[0] );
+		return inPlug()->metadataPlug()->getValue();
```

Comment:
It'd probably be a pretty small optimisation, but if the equivalent hash code path did `h = inPlug()->metadataPlug()->hash()` (and omitted the call to `ImageProcessor::hashMetadata()`, then we'd be sharing cache entries for the passed-through metadata. The same applies to the pass-through path for the format plug.